planning/002_file_upload_api.md

```diff
@@ -0,0 +1,74 @@
+    - the binary content of the new (different) file.
+4. For each failure, a client MAY provide:
+    - the content of the original file itself (to provide a side-by-side view)
+    - the path/filename where the new file was saved locally
+    - a list of additional files, for example a diff file, an image high contrast version, etc.. each must have:
+        - label (required): used in the UI to identify it (e.g. "high contrast", "diff") - we may have "special"
+        - file_name (required): path/name.ext
+        - content (required): binary content of the file
+5. For each file provided, an optional mimetype can be provided for better display in the UI (otherwise it'll be inferred. Any filetype is supported, but better UI will be provided for: "text/\*", "image/png", "image/svg+xml", "image/jpg", "image/webp", "image/gif")
+6. Each file will have a (configurable) maximum file size
+
+### API format
+
+There are a few options we might consider:
+
+1. Single REST endpoint with `multipart/form-data` content type, all files and metadata in a single request
+2. Single REST endpoint with `application/json` content type, files encoded as base64
+3. Multiple REST endpoints:
+    - POST `/api/v1/projects/<project_id>/` with `application/json` content type to create a test run.
+    - (optional) POST `/api/v1/projects/<project_id>/<test_run_id>/failures/` add a failure to the test run.
+    - POST `/api/v1/projects/<project_id>/test_runs/<test_run_id>/files/` with `multipart/form-data`, containing file bytes and metadata
+4. gRPC service, single rpc with client streaming (first message is metadata, then binary chunks follow)
+5. gRPC service, multiple rpcs:
+    ```protobuf
+    service Uploader {
+        rpc CreateTestRun(TestRunMeta) returns (CreateTestRunResponse);
+        (optional) rpc CreateFailure(FailureMetadata) returns (CreateFailureResponse);
+        rpc UploadFile(File) returns (UploadFileResponse);
+    }
+    ```
+6. Non-http communication (e.g. sftp/rsync)?
+7. Investigate what GCS/S3/other file storage services do
+
+Option 3 seems to be easily discountable, as it provides no benefits over the option 1 aside from ease
+of parsing/encoding (marginal, multipart should be well supported) and significant file-size
+increase
+
+Of the other REST options (1 and 3), the single multipart request is superior for performance
+when the connection is stable: everything can be streamed over a single connection
+The multiple requests option requires more round trips (and, if not careful, multiple http connections),
+but it significantly improves performance in case of failure, as only the failed operations
+need retrying.
+
+For the gRPC options similar considerations can be made, although connection reuse is built-in.
+
+Between gRPC and REST, I _think_ gRPC would make it easier to implement resumable uploads
```

Comment:
I personally don't have a lot of experience implementing gRPC services, but I do agree if feels suitable for our usecase. I'm happy to give it a go.